### PR TITLE
Compilation error regression in 4.1-RC1

### PR DESCRIPTION
**Pull request by wpopielarski**
Compilation error regression in 4.1-RC1: Cannot resolve imports to Java classes in some context.
Assembla ticket #1002456
Description:
There is pure java project A and depended scala project B.
Eclipse compiler doesn't compile java sources of A so that B is full of errors in places where depends on output of A.

### CODE REVIEW

File: org.scala-ide.sdt.aspects/src/scala/tools/eclipse/contribution/weaving/jdt/jcompiler/BuildManagerStore.java [hunk 1]
```diff
@@ -119,4 +126,16 @@ private IProject project(AbstractImageBuilder imageBuilder) {
 	}
 	return sourcesToCompile;
   }
+  
+  private boolean isScalaProject(IProject project) {
+	try {
+	  final String SCALA_NATURE_ID = "org.scala-ide.sdt.core.scalanature";
+	  if (project.hasNature(SCALA_NATURE_ID)) {
```
Comment (pkukielka) [inline org.scala-ide.sdt.aspects/src/scala/tools/eclipse/contribution/weaving/jdt/jcompiler/BuildManagerStore.java]: Just
`return project.hasNature(SCALA_NATURE_ID)`
would be enough ;)

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScalaProjectDependedOnJavaProjectTest.scala [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  val projectJName = "scalaDependedOnJavaJ"
+  val projectSName = "scalaDependedOnJavaS"
+  var projectJ: IJavaProject = _
+  var projectS: IScalaProject = _
+  val bundleName = "org.scala-ide.sdt.core.tests"
+
+  private val withSrcOutputStructure: SrcPathOutputEntry = (project, jProject) => {
+    val mainSourceFolder = project.getFolder("/src/main")
+    val mainOutputFolder = project.getFolder("/target/main")
+    val srcOuts = List(
```
Comment (pkukielka) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScalaProjectDependedOnJavaProjectTest.scala]: Cannot you just directly write:
```
JavaCore.newSourceEntry(
    jProject.getPackageFragmentRoot(mainSourceFolder).getPath,
    Array[IPath](),
    jProject.getPackageFragmentRoot(mainOutputFolder).getPath)
```
?

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScalaProjectDependedOnJavaProjectTest.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+object ScalaProjectDependedOnJavaProjectTest extends Before {
+  import org.scalaide.core.testsetup.SDTTestUtils._
+  val projectJName = "scalaDependedOnJavaJ"
+  val projectSName = "scalaDependedOnJavaS"
+  var projectJ: IJavaProject = _
+  var projectS: IScalaProject = _
+  val bundleName = "org.scala-ide.sdt.core.tests"
+
+  private val withSrcOutputStructure: SrcPathOutputEntry = (project, jProject) => {
```
Comment (pkukielka) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScalaProjectDependedOnJavaProjectTest.scala]: I would prefer def with explicit parameter types but I guess it depends on tastes ;)

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScalaProjectDependedOnJavaProjectTest.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  import Bdd._
+
+  @Test def shouldCorrectlyBuildScalaProjectWhichDependsOnJavaOne(): Unit = {
+    givenCleanWorkspaceForProjects(projectJ, projectS)
+
+    buildWorkspace()
+
+    val errors = markersMessages(findProjectProblemMarkers(projectS.underlying, SdtConstants.ProblemMarkerId).toList)
+
+    Assert.assertTrue("what's up?: " + errors.mkString(", "), errors.isEmpty)
```
Comment (pkukielka) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScalaProjectDependedOnJavaProjectTest.scala]: "what's up?: " 
? ^^

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+          val src = new File(sourceWorkspaceLoc(bundleName).toFile().getAbsolutePath + File.separatorChar + name)
+          val dst = new File(wspaceLoc.toFile().getAbsolutePath + File.separatorChar + name)
+          FileUtils.copyDirectory(src, dst)
+          done = true
+          cond.signal()
+        }
+      }
+    }
+    withLock(lock) {
+      while (!done) cond.await()
```
Comment (pkukielka) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala]: This won't work. You need somehow wait for for all threads. Maybe there is some way to do that in Eclipse API itself?

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala [hunk 6]
```diff
@@ -347,12 +352,26 @@ object SDTTestUtils extends HasLogger {
...
+      withSourceFolders(project, jProject) ++ Seq(JavaCore.newContainerEntry(Path.fromPortableString(SdtConstants.ScalaLibContId)))
+    def addScalaNature(project: IProject) = {
+      val description = project.getDescription
+      description.setNatureIds(description.getNatureIds ++ Array(SdtConstants.NatureId))
+      project.setDescription(description, null)
+      project
+    }
 
-    ScalaPlugin().getScalaProject(project)
+    ScalaPlugin().getScalaProject(addScalaNature(internalCreateJavaProjectInWorkspace(projectName, withScalaFolders).getProject))
```
Comment (pkukielka) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: You can use `createJavaProjectInWorkspace` method which you created bellow.

File: org.scala-ide.sdt.aspects/src/scala/tools/eclipse/contribution/weaving/jdt/jcompiler/BuildManagerStore.java [hunk 7]
```diff
@@ -81,42 +86,60 @@ public void setJavaSourceFilesToCompile(File[] files, IProject project) {
...
-	for (SourceFile source: sources) {
-	  File file = source.resource.getRawLocation().makeAbsolute().toFile();
-	  if (projectFiles.contains(file)) {
-		sourcesToCompile.add(source);
-	  }
-	}
-	return sourcesToCompile;
+    IProject project = project(imageBuilder);
+    if (!isScalaProject(project)) {
+      return new ArrayList<SourceFile>(sources);
```
Comment (dragos) [inline org.scala-ide.sdt.aspects/src/scala/tools/eclipse/contribution/weaving/jdt/jcompiler/BuildManagerStore.java]: Minor thing, but do we need to copy `sources` into a new array list? Couldn't we simply return `sources` here?

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScalaProjectDependedOnJavaProjectTest.scala (hunk 3 above)
Comment (dragos) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScalaProjectDependedOnJavaProjectTest.scala]: :+1:

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScalaProjectDependedOnJavaProjectTest.scala [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+      jProject.getPackageFragmentRoot(mainOutputFolder).getPath))
+  }
+
+  @BeforeClass def setup(): Unit = {
+    initializeProjects(bundleName, Seq(projectJName, projectSName)) {
+      projectJ = createJavaProjectInWorkspace(projectJName, withSrcOutputStructure)
+      projectS = createProjectInWorkspace(projectSName, withSrcOutputStructure)
+      addToClasspath(projectS, JavaCore.newProjectEntry(projectJ.getProject.getFullPath, false))
+      projectJ.getProject.refreshLocal(IResource.DEPTH_INFINITE, null)
+      projectS.underlying.refreshLocal(IResource.DEPTH_INFINITE, null)
```
Comment (dragos) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScalaProjectDependedOnJavaProjectTest.scala]: Are these refresh calls needed?

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScopeCompileTest.scala [hunk 9]
```diff
@@ -61,23 +62,13 @@ object ScopeCompileTest {
...
-        val dst = new File(wspaceLoc.toFile().getAbsolutePath + File.separatorChar + name)
-        FileUtils.copyDirectory(src, dst)
-        done = true
-      }
+    initializeProjects(bundleName, Seq(projectAName, projectBName)) {
+      projectA = createProjectInWorkspace(projectAName, withSrcOutputStructure)
+      projectB = createProjectInWorkspace(projectBName, withSrcOutputStructure)
+      addToClasspath(projectB, JavaCore.newProjectEntry(projectA.underlying.getFullPath, false))
+      projectA.underlying.refreshLocal(IResource.DEPTH_INFINITE, null)
+      projectB.underlying.refreshLocal(IResource.DEPTH_INFINITE, null)
```
Comment (dragos) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScopeCompileTest.scala]: The section above looks very similar to the one above, in `@BeforeClass def setup()`. Could they be refactored? I don't mind a bit of duplication if it's clearer this way, but I'm wondering if you noticed and it was intentinoal.

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    val cond = lock.newCondition()
+    projectNames.foreach { name =>
+      EclipseUtils.workspaceRunnableIn(workspace) { monitor =>
+        withLock(lock) {
+          val wspaceLoc = workspace.getRoot.getLocation
+          val src = new File(sourceWorkspaceLoc(bundleName).toFile().getAbsolutePath + File.separatorChar + name)
+          val dst = new File(wspaceLoc.toFile().getAbsolutePath + File.separatorChar + name)
+          FileUtils.copyDirectory(src, dst)
+          done = true
+          cond.signal()
```
Comment (dragos) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala]: This code looks fishy for several reasons.

- the code is copied from `org.scalaide.core.testsetup.SDTTestUtils.internalSetupProject(name: String, bundleName: String)`. Why? Can't we reuse the code there?
- the `workspaceRunnableIn` call could wrap the loop itself, so that all projects are copied in one go, instead of creating several transactions in a sequence. Is there a reason not to do that?
- the `await` call lower down will trigger after the first project in the list, but it doesn't seem to be the intention. Is that the intention?

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala (hunk 6 above)
Comment (wpopielarski) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: createJavaProjectInWorkspace is a public one and SDTTestUtils is a trait so is safer to create private method even if it is the whole body of public method

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScopeCompileTest.scala (hunk 9 above)
Comment (wpopielarski) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ScopeCompileTest.scala]: these parts are similar in fact but I treat them as a specific test description. In other test I can have let say 3 projects and for example no dependencies between them. I think it would be refactored maybe to some default configuration like 2 scala projects where B depends on A. Not a case here. I'm checking is 'refreshLocal' is really needed (I added it a little bit defensively)

File: .gitignore [hunk 11]
```diff
@@ -1,4 +1,4 @@
-**/.cache
+**/.cache*
```
Comment (sschaef) [inline .gitignore]: The commit with this change is obsolete, this has already been changed.

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+import SDTTestUtils.enableAutoBuild
+import SDTTestUtils.getErrorMessages
+import SDTTestUtils.sourceWorkspaceLoc
+import SDTTestUtils.workspace
+
+/**
+ * Behavioral Driven Development methods.
+ */
+trait Bdd {
```
Comment (sschaef) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala]: This name seems odd to me. Why not putting its member directly in `SDTTestUtils`?

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  def givenCleanWorkspaceForProjects(projects: IProject*): Unit = {
+    workspace.build(IncrementalProjectBuilder.CLEAN_BUILD, new NullProgressMonitor)
+    projects.foreach { project =>
+      Assert.assertTrue(getErrorMessages(project).isEmpty)
+    }
+  }
+}
+
+object Bdd {
+  implicit def jProjectToIProject(project: IJavaProject): IProject = project.getProject
```
Comment (sschaef) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala]: I'm not a fan of such implicit conversions, they are not worth of the few saved characters. Please remove them, the code is easier to understand without them.

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  implicit def sProjectToIProject(project: IScalaProject): IProject = project.underlying
+}
+
+/**
+ * Before actions for Scala IDE tests.
+ */
+trait Before {
```
Comment (sschaef) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala]: I don't think this type is required either. `SDTTestUtils` seems fine to me.

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala [hunk 15]
```diff
@@ -39,10 +39,11 @@ import org.eclipse.jdt.launching.JavaRuntime
 import org.scalaide.core.compiler.IScalaPresentationCompiler
 import org.scalaide.core.internal.project.ScalaProject
 
-/** Utility functions for setting up test projects.
+/**
+ * Utility functions for setting up test projects.
  *
  */
-object SDTTestUtils extends HasLogger {
+trait SDTTestUtils extends HasLogger {
```
Comment (sschaef) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: Why the change to a trait? Can there be/Do you want to have multiple instance of this type?

**Comment by dragos** [comment]
I tried out the binaries and it seemed to fix it for me. I didn't look at the code, though.

**Comment by sschaef** [comment]
[#1002461](https://scala-ide-portfolio.assembla.com/spaces/scala-ide/tickets/1002461#/activity/ticket:) is not a problem for me, the other ticket is fixed. Is there something special I have to do to reproduce the first ticket?

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala (hunk 15 above)
Comment (wpopielarski) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: it looks like a kind of util for integration test creation. when I look at scalatest or specs2 such utilities are traits usually not objects to import... I change it if it sores your eyes

Comment (sschaef) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: Full test frameworks have some other design goals to their code than what we have. Changing it to a trait means that it needs more Scaladoc because it needs to explain why and where it should instantiated.

Comment (wpopielarski) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: well im not going to die for my opinion but we definitely should do
somethhng with test utils. writing test now is let me say
discouraging. let talk on monday

2015-05-15 16:24 GMT+02:00, Simon Schäfer <notifications@github.com>:

Comment (sschaef) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: If there is something you don't like with the test suite, please change it and educate us. I just don't want to introduce new types everywhere without seeing a compelling reason.

Comment (wpopielarski) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: so the idea is following:
- let move a part which creates workspace to object
- the rest break up into functional parts:
  - ProjectUtils = project's creation, initialization and clean up
  - ErrorUtils = markers scanning on project, compile units
  - FileManipulationUtils = compile units, content replacement
  - maybe something else ?


2015-05-18 9:39 GMT+02:00 Simon Schäfer <notifications@github.com>:

Comment (sschaef) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: Ok. Do you want to have this now or can we move refactorings to post 4.1?

Comment (wpopielarski) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: yea, I revert this change and think how to do this seamless in post 4.1. No
need to delay solution delivery. And there is a need to educate as you
mentioned :).

2015-05-18 10:42 GMT+02:00 Simon Schäfer <notifications@github.com>:

Comment (wpopielarski) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: Wow, I missed SDTTestProject, maybe this guy should be extended?

2015-05-18 10:48 GMT+02:00 Wieslaw Popielarski <wpopielarski@gmail.com>:

Comment (wpopielarski) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: ok,
there are code duplications in SDTTestProject  and SDTTestUtils. I see
three ways of a project creation. It definitely needs some refactoring to
make life easier.
My idea for now is to leave Bdd as it is (if we haven't got a time, if we
have I can move it to SDTTestUtils), revert SDTTestUtils to object only and
finally open assembla ticket against test utils. Is it OK?

2015-05-18 11:09 GMT+02:00 Wieslaw Popielarski <wpopielarski@gmail.com>:

Comment (sschaef) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/SDTTestUtils.scala]: I don't like the name Bdd, it is non-descriptive. But sounds good otherwise.

File: org.scala-ide.sdt.aspects/src/scala/tools/eclipse/contribution/weaving/jdt/jcompiler/BuildManagerStore.java [hunk 16]
```diff
@@ -21,53 +23,56 @@
...
+
     // add the additional files
     IWorkspaceRoot workspaceRoot = ResourcesPlugin.getWorkspace().getRoot();
-    for (File file: files) {
-      for (IFile resource: workspaceRoot.findFilesForLocationURI(file.toURI())) {
-        // filter only the resources on the right project, to support nested projects
+    for (File file : files) {
+      for (IFile resource : workspaceRoot.findFilesForLocationURI(file.toURI())) {
+        // filter only the resources on the right project, to support nested
+        // projects
```
Comment (romanowski) [inline org.scala-ide.sdt.aspects/src/scala/tools/eclipse/contribution/weaving/jdt/jcompiler/BuildManagerStore.java]: Why move project to new line?

File: org.scala-ide.sdt.aspects/src/scala/tools/eclipse/contribution/weaving/jdt/jcompiler/BuildManagerStore.java [hunk 17]
```diff
@@ -81,42 +86,58 @@ public void setJavaSourceFilesToCompile(File[] files, IProject project) {
...
+        javaBuilderField = f;
+        javaBuilderField.setAccessible(true);
+      }
+    }
+    JavaBuilder builder = null;
+    try {
+      builder = (JavaBuilder) javaBuilderField.get(imageBuilder);
+      if (builder == null) {
+        throw new IllegalArgumentException(
+            "java builder of image builder is null");
```
Comment (romanowski) [inline org.scala-ide.sdt.aspects/src/scala/tools/eclipse/contribution/weaving/jdt/jcompiler/BuildManagerStore.java]: Why break this lines (also 106)

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala (hunk 13 above)
Comment (Kwestor) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala]: Ok, I checked usages of those and it seems sensible. Problem is we have several helper methods that requires `IProject`. Unfortunately both`IJavaProject` and `IScalaProject` encapsulates it, not extends.

With those implicits one can just pass `jProject: IJavaProject` to method that requires `IProject`.

One thing I would change here is rename `objecy Bdd` into `trait IProjectHelpers` and mix it to some base integration test.

File: .gitignore (hunk 11 above)
Comment (wpopielarski) [inline .gitignore]: what to do with it so? revert?

**Comment by sschaef** [comment]
LGTM
